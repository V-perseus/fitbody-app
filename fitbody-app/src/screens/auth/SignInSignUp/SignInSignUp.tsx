import React, { useState, useEffect, useRef, useLayoutEffect } from 'react'
import { View, Text, Keyboard, StatusBar, TouchableOpacity, Linking, ImageBackground, TextInput } from 'react-native'
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view'
import * as RNLocalize from 'react-native-localize'

// API
import api from '../../../services/api'

// Services & Data
import { updateSession } from '../../../services/session'
import { updateUser } from '../../../data/user'
import { checkExistingProgram } from '../../../data/workout'
import { getMonth, getWeek } from '../../../services/dates'
import { useSegmentLogger } from '../../../services/hooks/useSegmentLogger'

// Assets
import styles from './styles'
import globals from '../../../config/globals'
import BackgroundImage from '../../../../assets/images/signin/login-bg.png'
import Logo from '../../../../assets/images/logo/fitbody_white.svg'

// Components
import ContactSupport from '../../../components/ContactSupport/ContactSupport'
import { FormTextInput } from '../../../components/Forms/FormTextInput'
import { ButtonSquare } from '../../../components/Buttons/ButtonSquare'
import { HeaderButton } from '../../../components/Buttons/HeaderButton'

// Types
import { SignInSignUpOptionsProps } from '../../../config/routes/routeTypes'

const CURRENT_ENV = globals.currentEnv

interface ILoginProps extends SignInSignUpOptionsProps {}
const Login: React.FC<ILoginProps> = ({ navigation }) => {
  const { identifyUser } = useSegmentLogger()

  const [email, setEmail] = useState<string | null>(null)
  const [password, setPassword] = useState<string | null>(null)
  const [validating, setValidating] = useState(false)
  const [validEmail, setValidEmail] = useState(false)
  const [validPassword, setValidPassword] = useState(false)

  useEffect(() => {
    setValidEmail(email !== null && /(.+)@(.+){2,}\.(.+){2,}/.test(email))
    setValidPassword(password !== null && password.length > 1)
  }, [email, password])

  const emailInput = useRef<TextInput>(null)
  const passwordInput = useRef<TextInput>(null)

  useEffect(() => {
    getMonth()
    getWeek()
  }, [])

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    })
  }, [navigation])

  function signIn() {
    Keyboard.dismiss()
    setValidating(true)
    if (email && password && validEmail && validPassword) {
      api.users
        .signInWithEmail({
          email,
          password,
          timezone: RNLocalize.getTimeZone(),
          trial: true,
        })
        .then((data) => {
          // Error shown in api.js
          if (typeof data !== 'undefined') {
            identifyUser(data.user)
            // strip out events field coming from backend. It pollutes app state
            updateUser({
              ...data.user,
              events: [],
            })
            checkExistingProgram(data.user.workout_goal)
            updateSession({
              userId: data.user.id,
              apiToken: data.user.api_token,
              activePlan: data.active_plan,
            })
          }
        })
        .catch(() => {
          // don't need to do anything here... just catch the rejection
          // so the app doesn't crash lmao
        })
    }
  }

  return (
    <KeyboardAwareScrollView
      showsVerticalScrollIndicator={false}
      contentContainerStyle={{ marginBottom: 0 }}
      style={styles.scrollView}
      enableOnAndroid={true}
      keyboardShouldPersistTaps="handled"
      testID="login_scrollview"
      getTextInputRefs={() => {
        return [emailInput, passwordInput]
      }}>
      <StatusBar barStyle="light-content" />
      <ImageBackground source={BackgroundImage} resizeMode="cover" style={styles.backgroundImage}>
        <Logo style={styles.logo} />
      </ImageBackground>

      <HeaderButton onPress={navigation.goBack} iconColor={globals.styles.colors.colorWhite} style={styles.headerButton} />

      <View style={styles.view}>
        <Text style={styles.loginText}>LOG IN</Text>
        {/* Email */}
        <FormTextInput
          ref={emailInput}
          containerStyle={{ marginBottom: 10 }}
          label="Email:"
          value={email || ''}
          validating={validating}
          valid={validEmail}
          showValidation
          validationText="Please enter a valid email address"
          onChangeText={setEmail}
          autoFocus={true}
          keyboardType="email-address"
          returnKeyType="next"
          onSubmitEditing={() => passwordInput.current?.focus()}
          testID="login_email"
        />

        {/* Password */}
        <FormTextInput
          ref={passwordInput}
          label="Password:"
          valid={validPassword}
          validating={validating}
          validationText="Please enter your password"
          secureTextEntry={true}
          value={password || ''}
          returnKeyType="send"
          onChangeText={setPassword}
          onSubmitEditing={signIn}
          testID="login_password"
        />

        <View style={styles.forgotContainer}>
          <TouchableOpacity onPress={() => navigation.navigate('ForgotPassword')}>
            <Text style={styles.forgotText}>Forgot Password</Text>
          </TouchableOpacity>
        </View>

        <View style={styles.buttonContainer}>
          <ButtonSquare
            onPress={signIn}
            style={styles.signIn}
            text={`LOG IN ${CURRENT_ENV.includes('staging') ? 'TO STAGING' : ''}`}
            textStyle={styles.buttonText}
            testID="login_button"
          />
        </View>

        <ContactSupport
          style={{ marginVertical: 20 }}
          color={globals.styles.colors.colorBlack}
          linkColor={globals.styles.colors.colorLove}
          navigate={navigation.navigate}
          onPress={() => Linking.openURL('mailto:hello@fitbodyapp.com?subject=Question about my Fit Body app')}
        />
      </View>
    </KeyboardAwareScrollView>
  )
}

export default Login
