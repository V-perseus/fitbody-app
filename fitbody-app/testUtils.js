import React from 'react'
import { render as rtlRender } from '@testing-library/react-native'
import { configureStore } from '@reduxjs/toolkit'
import { NavigationContainer } from '@react-navigation/native'
import { createStackNavigator } from '@react-navigation/stack'
import { Provider } from 'react-redux'

// Import your own reducer
import { appReducer } from './src/store'
import { ModalStackScreens } from './src/config/routes/loggedInStack'

// custom render method override
function render(ui, { preloadedState, store = configureStore({ reducer: appReducer, preloadedState }), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react-native'
// override render method
export { render }

// custom navigation mock
const Stack = createStackNavigator()
export const MockedNavigator = ({
  component,
  params = {},
  preloadedState,
  store = configureStore({ reducer: appReducer, preloadedState }),
}) => {
  return (
    <Provider store={store}>
      <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen name="MockedScreen" component={component} initialParams={params} />
          <Stack.Screen name="Modals" component={ModalStackScreens} />
        </Stack.Navigator>
      </NavigationContainer>
    </Provider>
  )
}
