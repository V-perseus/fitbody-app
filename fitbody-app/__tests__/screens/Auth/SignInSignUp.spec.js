import SignInSignUp from '../../../src/screens/auth/SignInSignUp'
import { MockedNavigator, render, cleanup } from '../../../testUtils'

describe('SignInSignUp Screen', () => {
  let component = null
  beforeEach(() => {
    component = MockedNavigator({ component: SignInSignUp })
  })
  afterEach(() => {
    cleanup()
  })

  it('renders snapshot correctly', async () => {
    const tree = render(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Should render screen text', async () => {
    const { queryByText, queryAllByText } = render(component)
    expect(queryAllByText(/LOG IN/gi)).toHaveLength(2)
    expect(queryByText(/Email:/gi)).not.toBeNull()
    expect(queryByText(/Password:/gi)).not.toBeNull()
    expect(queryByText(/Forgot Password/gi)).not.toBeNull()
    expect(queryByText(/Have Questions\?/gi)).not.toBeNull()
  })
})
