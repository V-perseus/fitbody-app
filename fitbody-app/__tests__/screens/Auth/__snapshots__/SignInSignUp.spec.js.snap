// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInSignUp Screen renders snapshot correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            Object {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        accessibilityIgnoresInvertColors={true}
                        style={
                          Object {
                            "alignItems": "center",
                            "display": "flex",
                            "height": 724.6376811594203,
                            "justifyContent": "center",
                            "width": 750,
                          }
                        }
                      >
                        <Image
                          resizeMode="cover"
                          source="login-bg.jpg"
                          style={
                            Array [
                              Object {
                                "bottom": 0,
                                "left": 0,
                                "position": "absolute",
                                "right": 0,
                                "top": 0,
                              },
                              Object {
                                "height": 724.6376811594203,
                                "width": 750,
                              },
                              undefined,
                            ]
                          }
                        />
                        <fitbody_white.svg
                          style={
                            Object {
                              "marginTop": 100,
                            }
                          }
                        />
                      </View>
                      <View
                        accessible={true}
                        collapsable={false}
                        focusable={true}
                        hitSlop={16}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "transparent",
                              "opacity": 1,
                            },
                            Object {
                              "alignItems": "center",
                              "borderRadius": 30,
                              "display": "flex",
                              "height": 42,
                              "justifyContent": "center",
                              "marginHorizontal": 7,
                              "width": 42,
                            },
                            Object {
                              "left": 0,
                              "position": "absolute",
                              "top": 60,
                              "zIndex": 5,
                            },
                          ]
                        }
                      >
                        <arrow-left.svg
                          color="#ffffff"
                          testID="back_arrow"
                        />
                      </View>
                      <View
                        style={
                          Object {
                            "flex": 1,
                            "marginHorizontal": 24,
                            "marginVertical": 32,
                          }
                        }
                      >
                        <Text
                          style={
                            Object {
                              "fontFamily": "BebasNeue",
                              "fontSize": 30,
                              "marginBottom": 21,
                            }
                          }
                        >
                          LOG IN
                        </Text>
                        <View
                          style={
                            Array [
                              Object {
                                "position": "relative",
                              },
                              Object {
                                "marginBottom": 10,
                              },
                            ]
                          }
                        >
                          <Text
                            style={
                              Object {
                                "color": "#333333",
                                "fontFamily": "OpenSans-Bold",
                                "fontSize": 16,
                                "marginBottom": 2,
                              }
                            }
                          >
                            Email:
                          </Text>
                          <TextInput
                            autoCapitalize="none"
                            autoCorrect={false}
                            autoFocus={true}
                            keyboardType="email-address"
                            onChangeText={[Function]}
                            onSubmitEditing={[Function]}
                            placeholderTextColor="rgba(0, 0, 0, 0.5)"
                            returnKeyType="next"
                            secureTextEntry={false}
                            style={
                              Array [
                                Object {
                                  "borderRadius": 3,
                                  "borderWidth": 1,
                                  "color": "#333333",
                                  "fontSize": 16,
                                  "height": 40,
                                  "marginBottom": 4,
                                  "paddingLeft": 6,
                                },
                                undefined,
                              ]
                            }
                            testID="loginemail"
                            value=""
                          />
                          <View
                            style={
                              Object {
                                "height": 23,
                              }
                            }
                          />
                        </View>
                        <View
                          style={
                            Array [
                              Object {
                                "position": "relative",
                              },
                              undefined,
                            ]
                          }
                        >
                          <Text
                            style={
                              Object {
                                "color": "#333333",
                                "fontFamily": "OpenSans-Bold",
                                "fontSize": 16,
                                "marginBottom": 2,
                              }
                            }
                          >
                            Password:
                          </Text>
                          <TextInput
                            autoCapitalize="none"
                            autoCorrect={false}
                            autoFocus={false}
                            keyboardType="default"
                            onChangeText={[Function]}
                            onSubmitEditing={[Function]}
                            placeholderTextColor="rgba(0, 0, 0, 0.5)"
                            returnKeyType="send"
                            secureTextEntry={true}
                            style={
                              Array [
                                Object {
                                  "borderRadius": 3,
                                  "borderWidth": 1,
                                  "color": "#333333",
                                  "fontSize": 16,
                                  "height": 40,
                                  "marginBottom": 4,
                                  "paddingLeft": 6,
                                },
                                undefined,
                              ]
                            }
                            value=""
                          />
                          <View
                            style={
                              Object {
                                "height": 23,
                              }
                            }
                          />
                        </View>
                        <View
                          style={
                            Object {
                              "flex": 1,
                              "flexDirection": "row",
                              "justifyContent": "flex-end",
                            }
                          }
                        >
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            onClick={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "opacity": 1,
                              }
                            }
                          >
                            <Text
                              style={
                                Object {
                                  "color": "#ff6c8f",
                                  "fontFamily": "OpenSans-Semibold",
                                  "fontSize": 14,
                                }
                              }
                            >
                              Forgot Password
                            </Text>
                          </View>
                        </View>
                        <View
                          style={
                            Object {
                              "flexDirection": "row",
                              "marginTop": 70,
                            }
                          }
                        >
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            onBlur={[Function]}
                            onClick={[Function]}
                            onFocus={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "opacity": 1,
                                },
                                Object {
                                  "alignItems": "center",
                                  "borderRadius": 8,
                                  "height": 56,
                                  "justifyContent": "center",
                                  "width": 114,
                                },
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#ff6c8f",
                                  "flex": 1,
                                  "flexDirection": "row",
                                  "height": 55,
                                  "justifyContent": "center",
                                },
                              ]
                            }
                          >
                            <Text
                              style={
                                Array [
                                  Object {
                                    "color": "#ffffff",
                                    "fontSize": 16,
                                  },
                                  Object {
                                    "color": "#ffffff",
                                    "fontFamily": "OpenSans",
                                    "fontSize": 16,
                                    "fontWeight": "600",
                                    "letterSpacing": -0.39,
                                    "lineHeight": 22,
                                    "textAlign": "center",
                                  },
                                ]
                              }
                            >
                              LOG IN 
                            </Text>
                          </View>
                        </View>
                        <View
                          style={
                            Array [
                              Object {
                                "alignSelf": "stretch",
                                "flexDirection": "row",
                                "justifyContent": "center",
                              },
                              Object {
                                "marginVertical": 20,
                              },
                            ]
                          }
                        >
                          <Text
                            style={
                              Array [
                                Object {
                                  "color": "#ffffff",
                                  "fontFamily": "OpenSans",
                                  "fontSize": 12,
                                  "letterSpacing": 0.4,
                                },
                                Object {
                                  "color": "#333333",
                                },
                              ]
                            }
                          >
                            Have Questions? 
                          </Text>
                          <RNGestureHandlerButton
                            collapsable={false}
                            exclusive={true}
                            handlerTag={2}
                            handlerType="NativeViewGestureHandler"
                            onGestureEvent={[Function]}
                            onGestureHandlerEvent={[Function]}
                            onGestureHandlerStateChange={[Function]}
                            onHandlerStateChange={[Function]}
                            rippleColor={0}
                            touchSoundDisabled={false}
                          >
                            <View
                              accessible={true}
                              collapsable={false}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                            >
                              <Text
                                style={
                                  Array [
                                    Object {
                                      "color": "#ffffff",
                                      "fontFamily": "OpenSans-Bold",
                                      "fontSize": 12,
                                      "letterSpacing": 0.4,
                                    },
                                    Object {
                                      "color": "#ff6c8f",
                                    },
                                  ]
                                }
                              >
                                Contact Us.
                              </Text>
                            </View>
                          </RNGestureHandlerButton>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
